### PR TITLE
Refactoring scheduler + training

### PR DESCRIPTION
**Pull request by tchaton**
In this PR, schedulers will be refactored to handle more diversity and allow configuration directly from config file

[x] - Allow scheduler creation from config file (lr_scheduler, bn_scheduler)
[x] - Update model checkpoint
[x] - Test model checkpoint
[x] - Allow gradient accumulation
[] - Verify Plateau works with all metric_name

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @nicolas-chaulet , could you review this PR too ;)

File: src/core/schedulers/bn_schedulers.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            * bn_scheduler_params.bn_decay ** (int(it * batch_size / bn_scheduler_params.decay_step)),
+            bn_scheduler_params.bn_clip,
+        )
+    elif bn_scheduler_opt.bn_policy == "exponential_decay":
+        bn_lambda = lambda e: max(
+            eval(bn_scheduler_params.gamma) ** (e / bn_scheduler_params.decay_step),
+            bn_scheduler_params.bn_clip / bn_scheduler_params.bn_momentum,
+        )
+    else:
+        return NotImplementedError("learning rate policy [%s] is not implemented", bn_scheduler_opt.lr_policy)
```
Comment (nicolas-chaulet) [inline src/core/schedulers/bn_schedulers.py]: There are some lr scattered in this file here and there. Please clean.

File: src/core/schedulers/bn_schedulers.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    @property
+    def scheduler_opt(self):
+        return self._scheduler_opt
+
+    @scheduler_opt.setter
+    def scheduler_opt(self, scheduler_opt):
+        self._scheduler_opt = scheduler_opt
+
+    def step(self, epoch=None):
```
Comment (nicolas-chaulet) [inline src/core/schedulers/bn_schedulers.py]: Please add a test

File: src/core/schedulers/lr_schedulers.py [hunk 3]
```diff
@@ -1,46 +1,81 @@
...
 
 
 log = logging.getLogger(__name__)
 
+def repr(self, scheduler_params={}):
+    return "{}({})".format(self.__class__.__name__, scheduler_params)
 
-def build_basic_params():
-    return DictConfig({"base_lr": 0.001})
+class SchedulerWrapper():
```
Comment (nicolas-chaulet) [inline src/core/schedulers/lr_schedulers.py]: Rename to LRScheduler

File: src/metrics/model_checkpoint.py [hunk 4]
```diff
@@ -166,11 +218,12 @@ def save_best_models_under_current_metrics(self, model: BaseModel, metrics_holde
             model {[BaseModel]} -- [Model]
             metrics_holder {[Dict]} -- [Need to contain stage, epoch, current_metrics]
         """
-
         metrics = metrics_holder["current_metrics"]
         stage = metrics_holder["stage"]
         epoch = metrics_holder["epoch"]
 
+        model.set_metrics(metrics, stage, epoch)
```
Comment (nicolas-chaulet) [inline src/metrics/model_checkpoint.py]: Please remove

File: src/datasets/base_dataset.py [hunk 5]
```diff
@@ -55,6 +55,16 @@ def __init__(self, dataset_opt, training_opt):
...
                 setattr(self, new_name, transform)
 
+        self._dataset_state = None
+
+    @property
+    def dataset_state(self):
+        return self._dataset_state
+
+    @dataset_state.setter
+    def dataset_state(self, dataset_state):
```
Comment (nicolas-chaulet) [inline src/datasets/base_dataset.py]: I wonder why this is needed. It should not

File: src/metrics/model_checkpoint.py [hunk 6]
```diff
@@ -36,28 +40,34 @@ def get_model_checkpoint(
...
-        self._check_path = "{}.pt".format(check_name)
+        self._check_path = checkpoint_file
         self._initialize_objects()
 
     def _initialize_objects(self):
         self._objects = {}
         self._objects["models"] = {}
+        self._objects["model_state"] = None
         self._objects["stats"] = {"train": [], "test": [], "val": []}
         self._objects["optimizer"] = None
```
Comment (nicolas-chaulet) [inline src/metrics/model_checkpoint.py]: I think `lr_params` is not used anymore. Please remove

File: src/modules/KPConv/partial_dense.py [hunk 7]
```diff
@@ -57,3 +57,109 @@ def __init__(
         # PARAMTERS IMPORTANT FOR SHADOWING
         self.shadow_features_fill = 0.0
         self.shadow_points_fill_ = float(10e6)
+
+
+class KPConvPartialDense(BaseKPConvPartialDense):
```
Comment (nicolas-chaulet) [inline src/modules/KPConv/partial_dense.py]: Please remove

File: src/utils/config.py [hunk 8]
```diff
@@ -11,6 +11,13 @@
 
 from .enums import ConvolutionFormat
 
+def create_new_omega_conf(opt):
```
Comment (nicolas-chaulet) [inline src/utils/config.py]: What does this actually do?

File: src/utils/config.py [hunk 9]
```diff
@@ -11,6 +11,13 @@
...
+    empty_opt = OmegaConf.create({})
+    return OmegaConf.merge(opt, empty_opt)
+
+def merge_omega_conf(opt, d):
```
Comment (nicolas-chaulet) [inline src/utils/config.py]: Same, how is this different from a merge?

File: test/mockdatasets.py [hunk 10]
```diff
@@ -45,6 +46,13 @@ def set_strategies(self, model):
         transform = MultiScaleTransform(strategies)
         self._transform = transform
 
+    @property
+    def dataset_state(self):
```
Comment (nicolas-chaulet) [inline test/mockdatasets.py]: Make sure to update if you update the base class

File: test/test_model_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            shutil.rmtree(ckpt_dir)
+        except:
+            pass
+        if not os.path.exists(ckpt_dir):
+            os.makedirs(ckpt_dir)
+        name = "test"
+        model_checkpoint = ModelCheckpoint(ckpt_dir, name)
+        mock_metrics = {"current_metrics": {"acc": 12}, "stage": "test", "epoch": 10}
+        model_checkpoint.save_best_models_under_current_metrics(model, mock_metrics)
+        model_checkpoint.save_best_models_under_current_metrics(model, mock_metrics)
```
Comment (nicolas-chaulet) [inline test/test_model_checkpoint.py]: Please remove the extra one

File: test/test_model_checkpoint.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+        # Load checkpoint and initialize model
+        model_config = OmegaConf.merge(params, self.data_config)
+        model2 = _find_model_using_name(model_class, "segmentation", model_config, dataset)
+        model_checkpoint = ModelCheckpoint(ckpt_dir, name)
+        model_checkpoint.initialize_model(model2, weight_name="acc")
+        shutil.rmtree(ckpt_dir)
+
+        assert str(model.optimizer.__class__.__name__) == str(model2.optimizer.__class__.__name__)
```
Comment (nicolas-chaulet) [inline test/test_model_checkpoint.py]: use self.assert.. instead of the basic assert. It will be a bit clearer when things break

File: test/test_model_checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        model_checkpoint.save_best_models_under_current_metrics(model, mock_metrics)
+        model_checkpoint.save_best_models_under_current_metrics(model, mock_metrics)
+
+        model_checkpoint = ModelCheckpoint(ckpt_dir, name)
+        model = model_checkpoint.create_model_from_checkpoint(dataset)
+        print(model)
+        shutil.rmtree(ckpt_dir)
+
+    """
+    def test_create_model_from_checkpoint_without_dataset(self):
```
Comment (nicolas-chaulet) [inline test/test_model_checkpoint.py]: If not used please remove

File: src/utils/metrics_utils.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+
+def get_from_metric(latest_metrics, query_metric_name, current_stage):
```
Comment (nicolas-chaulet) [inline src/utils/metrics_utils.py]: What does this do? Please add a doctstring

File: train.py [hunk 15]
```diff
@@ -153,26 +152,22 @@ def main(cfg):
...
     model_config = OmegaConf.merge(model_config, cfg_training)
     model = instantiate_model(model_class, tested_task, model_config, dataset)
     log.info(model)
 
-    # Optimizer
-    model.set_optimizer(
-        getattr(torch.optim, otimizer_class, None), cfg_training.optimizer.params, cfg_training.learning_rate
-    )
+    # Initialize optimizer, schedulers
+    model.instantiate_optim(cfg)
```
Comment (nicolas-chaulet) [inline train.py]: Rename to instantiate_optimizer

File: src/models/base_model.py [hunk 16]
```diff
@@ -195,24 +294,30 @@ def search_from_key(modules):
...
             return search_with_keys(value, keys[1:], value_out)
+
         try:
             value_out = search_with_keys(opt, keys, value_out)
         except Exception as e:
-            print(e)
+            if msg_err:
+                raise Exception(str(msg_err))
+            else:
+                print(e)
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: Use `log.exception("Some message")` instead. It will print your message as well as the exception

File: src/models/base_model.py [hunk 17]
```diff
@@ -54,10 +80,25 @@ def lr_params(self):
...
             return None
 
+    @property
+    def schedulers(self):
+        return self._schedulers
+
+    @schedulers.setter
+    def schedulers(self, schedulers):
+        if schedulers:
+            self._schedulers = schedulers
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: I am wondering if you could have only one way to manage schedulers as opposed to these two ways. The setters and getters are here for that in principle, but at the moment the instantiate_optimi has to know that it should set those two variables when it should not. Maybe a method `set_cheduler(scheduler_name, scheduler)` would help you manage that in a cleaner way.

File: src/models/base_model.py [hunk 18]
```diff
@@ -90,17 +131,52 @@ def get_output(self):
     def forward(self) -> Any:
         """Run forward pass; called by both functions <optimize_parameters> and <test>."""
 
+    def set_selection_stage(self, set_selection_stage):
+        self._set_selection_stage = set_selection_stage
+
+    def set_metrics(self, metrics, stage, epoch):
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: I think we should remove that

File: src/models/base_model.py [hunk 19]
```diff
@@ -90,17 +131,52 @@ def get_output(self):
     def forward(self) -> Any:
         """Run forward pass; called by both functions <optimize_parameters> and <test>."""
 
+    def set_selection_stage(self, set_selection_stage):
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: Same I think this should not be there

File: src/models/base_model.py [hunk 20]
```diff
@@ -114,11 +190,32 @@ def get_current_losses(self):
...
+        optimizer_cls = getattr(torch.optim, optmizer_cls_name)
+        optimizer_params = {}
+        if hasattr(optimizer_opt, "params"):
+            optimizer_params = optimizer_opt.params
+        self._optimizer = optimizer_cls(self.parameters(), **optimizer_params)
+        colored_print(COLORS.Green, "Optimizer: {}".format(self._optimizer))
+
+        scheduler_opt = self.get_from_opt(config, ["training", "optim", "scheduler"])
+        if scheduler_opt:
+            self._lr_scheduler = instantiate_scheduler(self._optimizer, scheduler_opt)
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: See my comment in the schedulers setter.

File: src/models/base_model.py [hunk 21]
```diff
@@ -90,17 +131,52 @@ def get_output(self):
...
         """Calculate losses, gradients, and update network weights; called in every training iteration"""
         self._iterations += batch_size
         self.forward()  # first call forward to calculate intermediate results
-        self._optimizer.zero_grad()  # clear existing gradients
+        make_optimizer_step = self.manage_optimizer_zero_grad()  # Accumulate gradient if option is up
         self.backward()  # calculate gradients
         if self._grad_clip > 0:
             torch.nn.utils.clip_grad_norm_(self.parameters(), self._grad_clip)
-        self._optimizer.step()  # update parameters
+        if make_optimizer_step:
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: I think this whole logic should be tested. It is really dangerous and could break silently resulting in low performance training which would be really hard to debug.

File: src/metrics/model_checkpoint.py [hunk 22]
```diff
@@ -132,6 +165,25 @@ def __init__(
...
+        raise NotImplementedError
+
+    def create_model_from_checkpoint(self, dataset=None, weight_name=None):
+        if not self._checkpoint.is_empty:
+            model_state = self._checkpoint.get_model_state()
+            model_class = model_state["model_class"]
+            option = OmegaConf.create(model_state["option"])
+            dataset_state = model_state["dataset_state"]
+            task = dataset_state["task"]
+            if not dataset:
```
Comment (nicolas-chaulet) [inline src/metrics/model_checkpoint.py]: How would you be in this position? Is it in preparation for a forward only method? If yes then let's ignore it for now and revisit when we get there. A forward only script would still define a dataset anyway.

File: src/metrics/model_checkpoint.py [hunk 23]
```diff
@@ -132,6 +165,25 @@ def __init__(
...
 
+    def create_mock_dateset(self, dataset_state):
+        raise NotImplementedError
+
+    def create_model_from_checkpoint(self, dataset=None, weight_name=None):
+        if not self._checkpoint.is_empty:
+            model_state = self._checkpoint.get_model_state()
+            model_class = model_state["model_class"]
+            option = OmegaConf.create(model_state["option"])
+            dataset_state = model_state["dataset_state"]
```
Comment (nicolas-chaulet) [inline src/metrics/model_checkpoint.py]: I think I would like to remove this dataset_state for now and think about a better way of handling that. I don't think it is needed for now.

File: src/metrics/model_checkpoint.py [hunk 24]
```diff
@@ -213,4 +266,6 @@ def save_best_models_under_current_metrics(self, model: BaseModel, metrics_holde
                 current_stat[metric_name] = metric_value
                 current_stat["best_{}".format(metric_name)] = metric_value
 
-        self._checkpoint.save_objects(models_to_save, stage, current_stat, model.optimizer, model.lr_params, **kwargs)
+        self._checkpoint.save_objects(
+            models_to_save, model.model_state, stage, current_stat, model.optimizer, model.schedulers, **kwargs
```
Comment (nicolas-chaulet) [inline src/metrics/model_checkpoint.py]: It is a bit weird to have this second model_state (models to save are model states as well). Any chance this new property could come from the `state_dict` method? Like that you have one function that defines the state of the model and that's it.

File: conf/models/segmentation/kpconv.yaml [hunk 25]
```diff
@@ -22,29 +22,6 @@ models:
             nn: [64, 64, 64, 64, 64]
             dropout: 0.5
 
-    PDSimpleKPConv:
```
Comment (nicolas-chaulet) [inline conf/models/segmentation/kpconv.yaml]: You need that one for the tests

File: conf/training/default.yaml [hunk 26]
```diff
@@ -1,22 +1,27 @@
...
-            lr: 0.001
-    learning_rate:
-        scheduler_type: "step_decay"
-        base_lr: ${optimizer.params.lr}
-        lr_decay: 0.5
-        decay_step: 200000
-        lr_clip: 1e-5
+    optim:
+        base_lr: 0.001
+        accumulated_gradient: 10 # Accumulate gradient accumulated_gradient * batch_size
```
Comment (nicolas-chaulet) [inline conf/training/default.yaml]: Should this not be -1 by default?

File: src/__init__.py [hunk 27]
```diff
@@ -47,11 +51,17 @@ def instantiate_model(model_class, task, option, dataset: BaseDataset) -> BaseMo
...
         raise NotImplementedError(
             "In %s.py, there should be a subclass of BaseDataset with class name that matches %s in lowercase."
             % (model_module, class_name)
         )
-    return model(option, "dummy", dataset, modellib)
+
+    option_container = OmegaConf.to_container(option)
+    model = model_cls(option, "dummy", dataset, modellib)
+
+    model_state = {"model_class": model_class, "option": option_container, "model_module": model_module, "task": task}
```
Comment (nicolas-chaulet) [inline src/__init__.py]: As per previous comment did you consider avoiding this extra state? In addition to the state_dict provided by pytorch? In my opinion there should only be one state to store all information (the module state_dict seems like a good candidate)

File: src/metrics/model_checkpoint.py (hunk 24 above)
Comment (nicolas-chaulet) [reply src/metrics/model_checkpoint.py]: Don't you think this can be confusing?

File: src/models/base_model.py [hunk 28]
```diff
@@ -43,7 +47,27 @@ def __init__(self, opt):
...
+        self._accumulated_gradient_step = self.get_from_opt(opt, ["optim", "accumulated_gradient"])
+        if self._accumulated_gradient_step:
+            if self._accumulated_gradient_step > 1:
+                self._accumulated_gradient_count = 0
+                colored_print(COLORS.Green, "Accumulated option activated {}".format(self._accumulated_gradient_step))
+            else:
+                raise Exception("accumulated_gradient should be greater than 1")
+
+    @property
+    def model_state(self):
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: See my previous comment. I think this one should be merged with pytorch state_dict variable.

File: test/test_models.py [hunk 29]
```diff
@@ -71,28 +71,8 @@ def test_kpconv(self):
         model.forward()
         model.backward()
 
-    def test_kpconvpretransform(self):
```
Comment (nicolas-chaulet) [inline test/test_models.py]: Could you please add this test back?

File: src/utils/config.py [hunk 30]
```diff
@@ -8,9 +8,41 @@
...
+    selection_stage = getattr(cfg, "selection_stage", None)
+    if not selection_stage:
+        selection_stage = "val" if has_val_loader else "test"
+    else:
+        if not has_val_loader and selection_stage == "val": 
+            raise Exception("Selection stage should be: test")
+    return selection_stage
+
+def merge_omega_conf(opt: DictConfig, d: dict):
+    """This function allows to merge a OmegaConf DictConfig with a python dictionnary"""
```
Comment (nicolas-chaulet) [inline src/utils/config.py]: Nice thanks

File: src/models/base_model.py [hunk 31]
```diff
@@ -43,7 +47,34 @@ def __init__(self, opt):
...
```
Comment (nicolas-chaulet) [inline src/models/base_model.py]: Please remove those ones
